src/shogun/kernel/GaussianARDKernel.cpp

```diff
@@ -90,28 +75,19 @@ SGMatrix<float64_t> CGaussianARDKernel::get_parameter_gradient(
 
 	if (!strcmp(param->m_name, "weights"))
 	{
-		SGMatrix<float64_t> derivative=get_kernel_matrix();
-
+		SGMatrix<float64_t> derivative(num_lhs, num_rhs);
 		for (index_t j=0; j<num_lhs; j++)
 		{
+			SGVector<float64_t> avec=((CDotFeatures *)lhs)->get_computed_dot_feature_vector(j);
 			for (index_t k=0; k<num_rhs; k++)
 			{
-				SGVector<float64_t> avec=
-					((CDenseFeatures<float64_t>*) lhs)->get_feature_vector(j);
-				SGVector<float64_t> bvec=
-					((CDenseFeatures<float64_t>*) rhs)->get_feature_vector(k);
-
-				REQUIRE(avec.vlen==bvec.vlen, "Number of right and left hand "
-						"features must be the same\n");
-
-				float64_t element=compute(j,k);
-				float64_t product=CMath::pow((avec[index]-bvec[index]), 2)
-						*(m_weights[index]/m_width);
-
-				derivative(j,k)=-2*element*product;
+				SGVector<float64_t> bvec=((CDotFeatures *)rhs)->get_computed_dot_feature_vector(k);
+				bvec.scale(-1.0);
+				bvec.add(avec);
```

Comment:
add a negative version? mmmmH, we should add this. Can you ping Rahul
@lambday ?